dune/xt/la/container/common.hh

```diff
@@ -762,49 +837,63 @@ public:
   } // CommonSparseMatrix(...)
 
   template <int ROWS, int COLS>
-  explicit operator Dune::FieldMatrix<ScalarType, ROWS, COLS>() const
+  explicit operator std::unique_ptr<Dune::FieldMatrix<ScalarType, ROWS, COLS>>() const
   {
-    std::lock_guard<std::mutex> DUNE_UNUSED(lock)(mutex_);
     assert(ROWS == num_rows_ && COLS == num_cols_);
-    Dune::FieldMatrix<ScalarType, ROWS, COLS> ret(ScalarType(0));
+    auto ret = XT::Common::make_unique<Dune::FieldMatrix<ScalarType, ROWS, COLS>>(ScalarType(0));
     for (size_t rr = 0; rr < ROWS; ++rr)
       for (size_t kk = row_pointers_->operator[](rr); kk < row_pointers_->operator[](rr + 1); ++kk)
+        (*ret)[rr][column_indices_->operator[](kk)] = entries_->operator[](kk);
+    return std::move(ret);
+  }
+
+  explicit operator Dune::DynamicMatrix<ScalarType>() const
+  {
+    Dune::DynamicMatrix<ScalarType> ret(rows(), cols(), ScalarType(0));
+    for (size_t rr = 0; rr < rows(); ++rr)
+      for (size_t kk = row_pointers_->operator[](rr); kk < row_pointers_->operator[](rr + 1); ++kk)
         ret[rr][column_indices_->operator[](kk)] = entries_->operator[](kk);
     return ret;
   }
 
   ThisType& operator=(const ThisType& other)
   {
-    entries_ = other.entries_;
-    row_pointers_ = other.row_pointers_;
-    column_indices_ = other.column_indices_;
-    num_rows_ = other.num_rows_;
-    num_cols_ = other.num_cols_;
+    if (this != &other) {
+      row_pointers_ = other.row_pointers_;
+      column_indices_ = other.column_indices_;
+      num_rows_ = other.num_rows_;
+      num_cols_ = other.num_cols_;
+      entries_ = other.unshareable_ ? std::make_shared<EntriesVectorType>(*other.entries_) : other.entries_;
+      mutexes_ =
+          other.unshareable_ ? std::make_shared<std::vector<std::mutex>>(other.mutexes_->size()) : other.mutexes_;
+      unshareable_ = false;
+    }
     return *this;
   }
 
   /// \}
   /// \name Required by ContainerInterface.
   /// \{
-
   inline ThisType copy() const
   {
     ThisType ret(*this);
+    ret.entries_ = std::make_shared<EntriesVectorType>(*entries_);
```

Comment:
In the copy constructor, a shallow copy is done unless ``unshareable_ == true``, and the ``copy()`` method has to do a deep copy. If `` unshareable_ == true`` we will copy two times, so maybe I can add a check for that.